src/generic/datavgen.cpp

```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
 }
 
 
+void RowRange::Add(const unsigned int idx)
+{
+    RangeEntity *rng = NULL;
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    while (rngIdx < count)
+    {
+        rng = &m_ranges[rngIdx];
+
+        if (idx >= rng->from && rng->to >= idx)
+        {
+            // index already in range
+            return;
+        }
+
+        if (rng->from == idx + 1)
+        {
+            rng->from = idx;
+            CleanUp(rngIdx);
+            return;
+        }
+        if (rng->to == idx - 1)
+        {
+            rng->to = idx;
+            CleanUp(rngIdx);
+            return;
+        }
+
+        if (rng->from > idx + 1)
+        {
+            break;
+        }
+
+        rngIdx += 1;
+    }
+//    wxLogMessage("New Range: %d" , count);
+
+    RangeEntity newRange;
+    newRange.from = idx;
+    newRange.to = idx;
+    m_ranges.Insert(newRange, rngIdx);
+}
+
+void RowRange::RemoveFrom(const unsigned int idx)
+{
+    RangeEntity *rng = NULL;
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    while (rngIdx < count)
+    {
+        rng = &m_ranges[rngIdx];
+        if (rng->from >= idx)
+        {
+            m_ranges.RemoveAt(rngIdx);
+            count = m_ranges.GetCount();
+            continue;
+        }
+        if (rng->to >= idx)
+        {
+            rng->to = idx - 1;
+        }
+
+        rngIdx += 1;
+    }
+}
+
+
+void RowRange::CleanUp(int idx)
+{
+    RangeEntity *rng = NULL;
+    RangeEntity *prevRng = NULL;
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    if (idx > 0)
+    {
+        rngIdx = idx - 1;
+    }
+    while (rngIdx <= idx + 1 && rngIdx < count)
+    {
+        rng = &m_ranges[rngIdx];
+
+        if (prevRng != NULL && prevRng->to >= rng->to)
+        {
+            m_ranges.RemoveAt(rngIdx);
+            count = m_ranges.GetCount();
+            continue;
+        }
+
+        prevRng = rng;
+        rngIdx += 1;
+    }
+
+}
+
+bool RowRange::Has(const unsigned int idx)
+{
+    RangeEntity rng;
+    size_t count = m_ranges.GetCount();
+    for (size_t rngIdx = 0; rngIdx < count; rngIdx++)
+    {
+        rng = m_ranges[rngIdx];
```

Comment:
Here pointers are not used, but we copy the value (granted, a small one) unnecessarily, so a reference would still be better:

    const RangeEntity& rng = m_ranges[rngIdx];